---
title: "Targeted Sequencing"
author: Szi Kay Leung
date: "`r Sys.Date()`"
output:
  rmdformats::material:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: true
    highlight: tango
    code_folding: hide
    fig_caption: true
    cards: true



---

```{r setup, include=FALSE, results="hide"}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message = FALSE ,fig.width=15, fig.height= 11.5)

list_of_packages <- c("ggplot2", "scales", "reshape", "gridExtra", "grid", "dplyr","stringr","viridis","extrafont","tidyr","purr","DT")
req_packages <- list_of_packages[!(list_of_packages %in% installed.packages()[,"Package"])]
if(length(req_packages)) install.packages(req_packages, repo="http://cran.rstudio.com/")

suppressMessages(library(ggplot2))
suppressMessages(library(scales))
suppressMessages(library(reshape2))
suppressMessages(library(gridExtra))
suppressMessages(library(grid))
suppressMessages(library(plyr))
suppressMessages(library(dplyr))
suppressMessages(library(stringr)) 
suppressMessages(library(viridis)) 
suppressMessages(library(extrafont))
suppressMessages(library(tidyr))
suppressMessages(library(purrr))
suppressMessages(library(DT))
suppressMessages(library(plotly))
suppressMessages(library(pheatmap))
suppressMessages(library(devtools))
#suppressMessages(library(easyGgplot2))
suppressMessages(library(cowplot))
suppressMessages(library(tibble))
suppressMessages(library(splitstackshape)) #for splitting cells into multiple columns
suppressMessages(library(rjson)) # json files

#font_import()
# source function scripts
source("/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/Scripts/IsoSeq_Tg4510/Rmarkdown_Input.R")
source("/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/Scripts/IsoSeq_Tg4510/1_Transcriptome_Annotation/Isoseq3_QC/RunStats_Function.R")
```
# Aims
Analysis of First Targeted Sequencing Run on Sequel 1:\
1. On-Target Rate\
2. Raw Coverage\
  + Number of Transcripts per Gene identified\
3. Filtering by ToFu\
4. Number of Isoforms per Gene identified\
```{r}
# targetedphenotype 
targetedpheno <- read.csv("/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/Scripts/IsoSeq_Tg4510/Raw_Data/Targeted_Transcriptome/Targeted_Sample_Demographics.csv") 
```

# Iso-Seq

```{r}
# Input IsoSeq Paths
IsoSeq_input_dir <- "/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/IsoSeq/Targeted_Transcriptome/Mouse/IsoSeq"
PostIsoSeq_input_dir <- "/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/IsoSeq/Targeted_Transcriptome/Mouse/Post_IsoSeq"

## Input CCS, LIMA, REFINE files
# CCS, LIMA summary
CCS <- read.csv(paste0(IsoSeq_input_dir, "/CCS/AllMouseTargeted_CCS_output.csv"), header = T)
LIMA <- read.csv(paste0(IsoSeq_input_dir, "/LIMA/AllMouseTargeted_LIMA_summary.csv"), header = T)
# REFINE summary input
REFINE_json_list <- list.files(paste0(IsoSeq_input_dir,"/REFINE/"), pattern = "flnc.filter_summary.json", full.names = T) %>% .[!grepl("Targeted_Seq",.)]
REFINE_list <- lapply(REFINE_json_list , function(x) as.data.frame(fromJSON(file = x)))
names(REFINE_list) <- list.files(paste0(REFINE_input_dir), pattern = "flnc.filter_summary.json") %>% .[!grepl("Targeted_Seq",.)]
for(i in 1:length(names(REFINE_list))){REFINE_list[[i]]$Sample <- names(REFINE_list)[[i]]}
# CLUSTER files 
# do not include batched report Targeted_Seq_1.clustered.cluster_report.csv
CLUSTER_input_files <- list.files(paste0(IsoSeq_input_dir,"/CLUSTER"), pattern = "clustered.cluster_report.csv", full.names = T) %>% .[!grepl("Targeted_Seq",.)]
CLUSTER_list <- lapply(CLUSTER_input_files, function(x) read.csv(x))
names(CLUSTER_list) <- list.files(paste0(CLUSTER_input_dir), pattern = "clustered.cluster_report.csv") %>% .[!grepl("Targeted_Seq",.)]
for(i in 1:length(names(CLUSTER_list))){CLUSTER_list[[i]]$Sample <- names(CLUSTER_list)[[i]]}
# MERGED CLUSTER FILES
Merged_CLUSTER <- read.csv(paste0(IsoSeq_input_dir,"/MERGED_CLUSTER/AllMouseTargeted.clustered.cluster_report.csv"))

# MAPPING DIR for probes
Mapping_dir <- paste0(PostIsoSeq_input_dir,"/MAP/TARGETRATE")
Merged_Mapping_dir <- paste0(PostIsoSeq_input_dir,"/MAP")
#Merged_probes_files <- read.table(paste0(Merged_Mapping_dir,"All_Targeted_Merged.fasta.sam.probe_hit.txt"),header=T, as.is=T, sep="\t")

# Apply function to concatenate all the reads
Reads <- number_of_reads()


Reads <- Reads %>% mutate(Batch = word(Reads$variable, c(3), sep = fixed("_"))) %>% 
  full_join(., targetedpheno, by = c("sample" = "Sample")) %>%
  unite("Batch", Batch.x,Batch.y, na.rm = TRUE) %>%
  mutate(Batch = recode(Batch, "3b" = "3", "3a" = "3 (partial run)")) %>%
  select(Description,value,sample,Batch) %>%
  bind_rows(data.frame(Description = "Clustered reads(Transcripts)",value = as.numeric(as.character(nrow(Merged_CLUSTER))),
           sample = "Merged", Batch = "Merged"))

pIsoSeq <- ggplot(Reads, aes(x = Description, y = value, colour = Batch, group = Batch)) +
      geom_line() + geom_point() +  mytheme + theme(legend.position = "bottom") + labs(x = "", y = "Number of Reads (Thousands)") +
  scale_y_continuous(labels = unit_format(unit = "", scale = 1e-3)) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))
  
pIsoSeq
```

# On-Target Rate

Using L.Tseng's script *(https://github.com/Magdoll/cDNA_Cupcake/wiki/Targeted-Iso-Seq-QC-Wiki)* to calculate the on-target rate from polished.fa file, SAM alignment, and probe BED file.

Batch 2 and Batch 3 had more samples (n = 9) than batch 6, therefore increased propensity to sequence targeted genes

**Note**: This is prior to Tofu-collapse, therefore some of different FL transcripts can still be derived from the same isoform

## {.tabset .tabset-pills}

### Individual

```{r}
Probes_input <- list.files(path = Mapping_dir, pattern = "fasta.sam.probe_hit.txt", full.names = T)
Probes_files <- lapply(Probes_input, function(x) read.table(x, header=T, as.is=T, sep="\t"))
names(Probes_files ) <- list.files(path = Mapping_dir, pattern = "fasta.sam.probe_hit.txt")

Probes <- merge(ldply(Probes_files , function(x) nrow(x)),
      ldply(Probes_files , function(x) length(which(x$num_base_overlap != "0"))),by = ".id") %>%
  `colnames<-`(c("file", "Total_mapped_reads", "reads_probe_hit")) %>% 
  mutate(perc = reads_probe_hit/Total_mapped_reads * 100) %>%
  mutate(sample = word(.$file, c(1), sep = fixed("."))) %>% 
  full_join(., targetedpheno, by = c("sample" = "Sample"))

pProbes<- ggplot(Probes, aes(x = as.factor(Batch), y = perc, fill = as.factor(Phenotype))) + geom_boxplot() +
  geom_point(aes(colour = as.factor(Phenotype)), position = position_jitterdodge()) + 
  mytheme + labs(y = "On-Target Rate", x = "Batch") + 
  scale_fill_manual(values = c(alpha(label_colour("TG"),0.4),alpha(label_colour("WT"),0.4)), name = "Phenotype") + 
  scale_colour_manual(values = c(label_colour("TG"),label_colour("WT")), guide=FALSE) 

pProbes


```

### Merged

```{r}
#sum(Reads[Reads$Description == "Clustered reads(Transcripts)","value"])
#nrow(Merged_CLUSTER)
```
A total of **`r dim(Merged_probes_files)[1]`** FL transcripts identified. Only an on-target rate of **`r round(length(which(Merged_probes_files$num_base_overlap != "0"))/dim(Merged_probes_files)[1]*100,0)`**%, i.e. % of FL transcripts with at least one of its bases matching to any primer. 

# Mapping

## {.tabset .tabset-pills}

### All Transcripts (including non-Targeted)

```{r}

Merged_mapping <- read.table(paste0(PostIsoSeq_input_dir,"/MAP/PAF/AllMouseTargeted_reads_with_alignment_statistics.txt"))
pMergedMap <- plot_mapping_quality(Merged_mapping)

pMergedMap 
```

### Targeted transcripts 

```{r}
#merged_targeted_transcripts <- Merged_probes_files %>% filter(num_base_overlap != "0") %>% select(read_id)
#plot_mapping_quality(Merged_mapping %>% filter(V1 %in% merged_targeted_transcripts$read_id))

```


# Target Genes (Merged data)

merged data
clustering, cupcake collapse, cupcake filtering, sqanti filtering
filter to only relevant genes, and number of isoforms

## {.tabset .tabset-pills}

### Filtering via pipeline
```{r}
All_Targeted_Merged <- "/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/IsoSeq/Targeted_Transcriptome/All_Targeted_Merged"
Merged_PostIso <- list(read.csv(paste0(All_Targeted_Merged,"/CLUSTER/All_Targeted_Merged.clustered.cluster_report.csv")), # Cluster
                       read.table(paste0(All_Targeted_Merged,"/TOFU/All_Targeted_Merged.collapsed.abundance.txt"), header = T), # Cupcake collapse
                       read.table(paste0(All_Targeted_Merged,"/TOFU/All_Targeted_Merged.collapsed.read_stat.txt"), header = T), 
                       read.table(paste0(All_Targeted_Merged,"/TOFU/All_Targeted_Merged.collapsed.filtered.abundance.txt"), as.is = T, sep = "\t", header = T), 
                       read.table(paste0(All_Targeted_Merged,"/SQANTI/All_Targeted_Merged.collapsed.filtered_classification.filtered_lite_classification.txt"),
                                  as.is = T, sep = "\t", header = T)
                       )

names(Merged_PostIso) <- c("cluster","cupcake","cupcake_readstat","cupcake_filter","sqanti_filter")


colnames(Merged_probes_files)[6] <- "Probes"
Merged_probes_files$Gene <- word(Merged_probes_files$Probes,c(3),  sep = fixed ('_'))
Merged_probes_files$Gene <- word(Merged_probes_files$Gene,c(1),  sep = fixed ('('))
Merged_probes_files$Gene[Merged_probes_files$Gene == "27642461"] <- "MAPT"
Merged_probes_files$Gene[Merged_probes_files$Gene == "27642460"] <- "ANK1"

# Filter by targeted genes
# targeted_transcripts = transcript/x of targeted genes (differented by having overlap with probes)
# targeted_read_id = mx/x/ccs of targeted genes (origianal reads)
# targeted_pbid = pbid of targeted genes (from readstat file after filtering for targeted_read_id)

targeted <- Merged_probes_files %>% filter(Merged_probes_files$num_base_overlap != "0") %>% select(read_id,Gene) %>% rename(transcript_id = read_id) %>%
  left_join(.,Merged_PostIso$cluster, by = c("transcript_id"= "cluster_id")) %>% 
  left_join(.,Merged_PostIso$cupcake_readstat, by = c("read_id"="id") )

# Example of unmapped transcript but probed
#Merged_probes_files %>% filter(read_id == "transcript/24")
#Merged_PostIso$cluster %>% filter(cluster_id == "transcript/24")
#Merged_PostIso$cupcake_readstat %>% filter(id == "m54082_200731_163617/50594276/ccs")

tgcluster <- targeted %>% group_by(transcript_id,Gene) %>% tally() %>% # multiple same transcript_id from clustering ccs reads 
  group_by(Gene) %>% tally() %>% mutate(type = "Clustered")

# PBids = NA for non-targeted genes
tgcupcake <- Merged_PostIso$cupcake %>% left_join(., unique(targeted[,c("pbid","Gene")])) %>% group_by(Gene) %>% tally() %>% filter(!is.na(Gene)) %>% mutate(type = "Cupcake")
tgcupcakefilter <- Merged_PostIso$cupcake_filter %>% left_join(., unique(targeted[,c("pbid","Gene")], by = "pbid")) %>% group_by(Gene) %>% tally() %>% filter(!is.na(Gene))  %>% mutate(type = "Cupcake Filter")

# if use probes, also misannotation of other genes nearby "Gm49227"
#tgsqfilter <- Merged_PostIso$sqanti_filter %>% left_join(., unique(targeted[,c("pbid","Gene")]), by = c("isoform" = "pbid")) %>% group_by(Gene) %>% tally() %>% filter(!is.na(Gene)) 
tgsqfilter <- Merged_PostIso$sqanti_filter %>% filter(toupper(associated_gene) %in% targeted$Gene) %>% group_by(associated_gene) %>% tally() %>% mutate(type = "Sqanti filter") %>% mutate(Gene = toupper(associated_gene))

pNumfil <- bind_rows(tgcluster, tgcupcake, tgcupcakefilter, tgsqfilter) %>%
  ggplot(., aes(x = Gene, y = n, fill = type)) + geom_bar(stat = "identity") +
  mytheme + labs(x = "", y = "Number of Transcripts (Thousand)") + theme(legend.position = c(0.8,0.8)) + 
  scale_fill_discrete(name = "", labels = c("Iso-Seq Cluster","Cupcake Collapse","Cupcake Filter","SQANTI filter")) + 
  scale_y_continuous(labels = ks) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

pNumfil

```

### Final dataset (SQANTI) {.tabset .tabset-pills}

#### Novel vs Annotated

```{r}
plot_sq_fil <- function(sq_df){
  p <- sq_df %>% filter(toupper(associated_gene) %in% targeted$Gene) %>% 
    select(isoform, structural_category, associated_gene, associated_transcript, subcategory) %>% 
    mutate(transcript_type = ifelse(.$associated_transcript == "novel","Novel","Annotated")) %>% 
    mutate(transcript_type = factor(transcript_type, levels = c("Novel","Annotated"))) %>%
    ggplot(., aes(x = associated_gene, fill = transcript_type)) + geom_bar() + 
    mytheme + labs(x = "", y = "Number of Isoforms") + 
    scale_fill_discrete(name = "Transcript Classification") + 
    theme(legend.position = c(0.9,0.9)) +
    theme(axis.text.x = element_text(angle = 45, hjust=1))
  
  return(p)
}

pNum1 <- plot_sq_fil(Merged_PostIso$sqanti_filter)
pNum1
```

#### Structural Category

```{r}

plot_structural_category <- function(sq_df){
  p <- sq_df %>% filter(toupper(associated_gene) %in% targeted$Gene) %>% 
    mutate(structural_category = factor(structural_category, levels = c("novel_not_in_catalog","novel_in_catalog",
                                                                       "incomplete-splice_match","full-splice_match"))) %>%
    ggplot(., aes(x = associated_gene, fill = structural_category)) + geom_bar() + 
    mytheme + labs(x = "", y = "Number of Isoforms") +
    theme(legend.position = c(0.9,0.9)) + 
    scale_fill_manual(name = "Transcript Classification", labels = c("NNC","NIC","ISM","FSM"),
                        values = c(alpha("#F8766D",0.8),alpha("#F8766D",0.3),alpha("#00BFC4",0.8),alpha("#00BFC4",0.3))) +
    theme(axis.text.x = element_text(angle = 45, hjust=1))
  
  return(p)
}

pNum2 <- plot_structural_category(Merged_PostIso$sqanti_filter)
pNum2
```

#### Subcategory

```{r}
plot_subcategory <- function(df_sq, structural_category_input){
  p <- df_sq %>% filter(toupper(associated_gene) %in% targeted$Gene) %>%
    select(associated_gene,structural_category,subcategory) %>%
    filter(structural_category == structural_category_input) %>%
    ggplot(., aes(x = associated_gene, fill = subcategory)) + geom_bar() + mytheme + 
    labs(y = "Number of Transcripts", x = "") +
    theme(axis.text.x = element_text(angle = 45, hjust=1),legend.position = "bottom")
    
  if(structural_category_input == "incomplete-splice_match"){
  p <-  p + scale_fill_discrete(labels = c("3' Fragment", "5' Fragment", "Internal Fragment","IR"))
  }
  return(p)
}

plot_subcategory(Merged_PostIso$sqanti_filter, "incomplete-splice_match") + labs(fill = "ISM Subcategory")
plot_subcategory(Merged_PostIso$sqanti_filter, "novel_in_catalog")
plot_subcategory(Merged_PostIso$sqanti_filter, "novel_not_in_catalog")  
```

# Batch 1

```{r}
B1sq <- read.table("/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/IsoSeq/Targeted_Transcriptome/Post_IsoSeq/SQANTI/Targeted_Seq_1/Targeted_Seq_1.collapsed.filtered_classification.filtered_lite_classification.txt", header = T)

B1sq %>% filter(toupper(associated_gene) %in% targeted$Gene) %>%
  mutate(transcript_type = ifelse(.$associated_transcript == "novel","Novel","Annotated")) %>% 
  mutate(transcript_type = factor(transcript_type, levels = c("Novel","Annotated"))) %>%
  ggplot(., aes(x = associated_gene, fill = transcript_type)) + geom_bar() + 
  mytheme + labs(x = "", y = "Number of Isoforms") + 
  scale_fill_discrete(name = "Transcript Classification") + 
  theme(legend.position = c(0.9,0.9)) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

# PB.1357.3: clusterin known transcript
B1sq %>% filter(associated_gene == "Cd33")
B1sq %>% filter(isoform == "PB.1357.3") 
B1readstat <- read.table("/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/IsoSeq/Targeted_Transcriptome/Post_IsoSeq/TOFU/Targeted_Seq_1.collapsed.read_stat.txt", header = T)
B1readstat %>% filter(pbid == "PB.1357.3")
Merged_PostIso$cupcake_readstat %>% filter(id == "m54082_191116_131337/47644765/ccs")
Merged_PostIso$cupcake_readstat %>% filter(id == "m54082_191116_131337/57737675/ccs")
Merged_PostIso$cupcake_filter %>% filter(pbid == "PB.7402.33")
Merged_PostIso$sqanti_filter %>% filter(isoform == "PB.7402.33")
Merged_PostIso$cupcake_filter %>% filter(pbid == "PB.7402.2")

```


# Alternative Pipeline

## {.tabset .tabset-pills}

### Pipeline

![Pipeline for analysis](/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/Scripts/IsoSeq_Tg4510/Images/All_Demux.png)

### Default settings (99% Coverage) {.tabset .tabset-pills}


```{r}
# after TAMA filtering and subset of SQANTI filtered dataset, generated with same pipeline (but added TAMA filtering at the end) and using 99% coverage threshold for cupcake 
Merged_PostIso_alt <- list(
  read.table(paste0(All_Targeted_Merged,"/Alternative_Pipeline/SQANTI/All_Targeted_Merged.collapsed_classification.filtered_lite_classification.txt"), header = T),
  read.table(paste0(All_Targeted_Merged,"/Alternative_Pipeline/SQANTI_TAMA_FILTER/All_Targeted_Merged_sqantitamafiltered.classification.txt"), header = T)
  )
names(Merged_PostIso_alt) <- c("sqanti_filter", "sqanti_tama_filter")
```

#### SQANTI Filter  {.tabset .tabset-pills}

##### Novel vs Annotated

```{r}
pNum1_sq_alt <- plot_sq_fil(Merged_PostIso_alt$sqanti_filter)
pNum1_sq_alt
```

##### Structural Category

```{r}
pNum2_sq_alt <- plot_structural_category(Merged_PostIso_alt$sqanti_filter)
pNum2_sq_alt
```

##### Subcategory

```{r}
pNum3_sq_alt <- plot_subcategory(Merged_PostIso_alt$sqanti_filter, "incomplete-splice_match") + labs(fill = "ISM Subcategory")
pNum3_sq_alt
```


#### TAMA Filter

##### Novel vs Annotated  {.tabset .tabset-pills}

```{r}
pNum1_tama_alt <- plot_sq_fil(Merged_PostIso_alt$sqanti_tama_filter)
pNum1_tama_alt
```

##### Structural Category

```{r}
pNum2_tama_alt <- plot_structural_category(Merged_PostIso_alt$sqanti_tama_filter)
pNum2_tama_alt
```

##### Subcategory

```{r}
pNum3_tama_alt <- plot_subcategory(Merged_PostIso_alt$sqanti_tama_filter, "incomplete-splice_match") + labs(fill = "ISM Subcategory")
pNum3_tama_alt
```
```{r}
pdf(paste0(output_plot_dir,"/TargetedTg4510_RunStats.pdf"), width = 15, height = 11.5)
plot_grid(pNumfil,pNum1, labels = c("a","b"), label_size = 25, label_fontfamily = "CM Roman", label_fontface = "bold",scale = 0.95, ncol = 1)
plot_grid(pNum1_sq_alt,pNum1_tama_alt, labels = c("a","b"), label_size = 25, label_fontfamily = "CM Roman", label_fontface = "bold",scale = 0.95, ncol = 1)
plot_grid(pNum2_sq_alt,pNum2_tama_alt, labels = c("a","b"), label_size = 25, label_fontfamily = "CM Roman", label_fontface = "bold",scale = 0.95, ncol = 1)
plot_grid(pNum3_sq_alt,pNum3_tama_alt, labels = c("a","b"), label_size = 25, label_fontfamily = "CM Roman", label_fontface = "bold",scale = 0.95, ncol = 1)
dev.off()
```


