---
title: "PacBio Tg4510 Control vs Tg4510 WT at 2 months"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir ="~")
```
## Input TG mouse long reads
```{r message = FALSE}
# Input: Isoseq O18 processed from Count.O18.Rmd  
# O18_PacBio_complete <- read.csv("//isad.isadroot.ex.ac.uk/UOE/User/My Documents/PacBio/AD_Mouse_Tg4510/2 months/O18/Isoseq3/O18_ordered_unique.csv", header = TRUE)[,-1]
O18_PacBio_complete <- read.csv("/mnt/data1/Szi/O18/Comparisons_RNAseq/O18_ordered_unique.csv", header = TRUE)[,-1]

head(O18_PacBio_complete)
# check all labelled genes include novel genes 
O18_PacBio_complete[which(is.na(O18_PacBio_complete$mgi_symbol)),]
library(data.table)
head(O18_PacBio_complete[which(O18_PacBio_complete$mgi_symbol %like% "novelGene"),])


# Only take columns forward for comparison with RNA-seq data: 1) gene names 2) PacBio ID for identification of different isoforms per gene 3) Number of FL counts 4) length for exact match of isoforms as associated gene and transcript can be the same
PacBio_O18 <- O18_PacBio_complete[,c("mgi_symbol","isoform","count_fl","length","associated_gene","associated_transcript")]
# As example: different isoforms for same gene 
PacBio_O18[which(PacBio_O18 == "Arf3"),]
O18_PacBio_complete[which(O18_PacBio_complete$mgi_symbol == "Cpe"),]
```
## Input WT mouse long reads
```{r}
# Input: Isoseq Q21 processed from Count.Q21.Rmd  
# Q21_PacBio_complete <- read.csv("//isad.isadroot.ex.ac.uk/UOE/User/My Documents/PacBio/AD_Mouse_Tg4510/2 months/Q21/Isoseq3/Q21_ordered_unique.csv", header = TRUE)[,-1]
Q21_PacBio_complete <- read.csv("/mnt/data1/Szi/Q21/Comparisons_RNAseq/Q21_ordered_unique.csv", header = TRUE)[,-1]

# Interestingly in merging, novel genes did not get annotated as novel  
Q21_PacBio_complete[which(is.na(Q21_PacBio_complete$mgi_symbol)),]
# For Gene column with novel genes, replace NA with associated gene 
# first to convert both columns into character 
Q21_PacBio_complete$mgi_symbol <- as.character(Q21_PacBio_complete$mgi_symbol)
Q21_PacBio_complete$associated_gene <- as.character(Q21_PacBio_complete$associated_gene)
# if gene column is NA, replace from text in associated gene column in respective row
Q21_PacBio_complete$mgi_symbol <- ifelse(is.na(Q21_PacBio_complete$mgi_symbol) == TRUE, Q21_PacBio_complete$associated_gene, Q21_PacBio_complete$mgi_symbol)


# Only take 3 columns forwarded for comparison with RNA-seq data: 1) gene names 2) PacBio ID for identification of different isoforms per gene 3) Number of FL counts 
PacBio_Q21 <- Q21_PacBio_complete[,c("mgi_symbol","isoform","count_fl","length","associated_gene","associated_transcript")]
# As example: different isoforms for same gene 
PacBio_Q21[which(PacBio_Q21 == "Arf3"),]
Q21_PacBio_complete[which(Q21_PacBio_complete$mgi_symbol == "Cpe"),]
```
## Merge into one df for Correlation 
```{r}
PB_2mons <-merge(PacBio_O18,PacBio_Q21, all = TRUE) 
```