---
title: "Targeted Sequencing 1: Per Individual Sample"
author: Szi Kay Leung
date: "`r Sys.Date()`"
knit: (function(inputFile, encoding) { 
      rmarkdown::render(inputFile,
                        encoding=encoding, 
                        output_file=paste0("/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/Targeted/Rmarkdowns",
                        '/Targeted_Sequencing_1_individual.html')) })
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE, results="hide"}
knitr::opts_chunk$set(echo = TRUE,fig.width=14, fig.height= 7)

list_of_packages <- c("ggplot2", "scales", "reshape", "gridExtra", "grid", "dplyr","stringr","viridis","extrafont","tidyr","purr","DT")
req_packages <- list_of_packages[!(list_of_packages %in% installed.packages()[,"Package"])]
if(length(req_packages)) install.packages(req_packages, repo="http://cran.rstudio.com/")

suppressMessages(library(ggplot2))
suppressMessages(library(scales))
suppressMessages(library(reshape2))
suppressMessages(library(gridExtra))
suppressMessages(library(grid))
suppressMessages(library(dplyr))
suppressMessages(library(stringr)) 
suppressMessages(library(viridis)) 
suppressMessages(library(extrafont))
suppressMessages(library(tidyr))
suppressMessages(library(purrr))
suppressMessages(library(DT))
suppressMessages(library(plotly))
suppressMessages(library(pheatmap))
suppressMessages(library(devtools))
suppressMessages(library(easyGgplot2))
suppressMessages(library(splitstackshape)) #for splitting cells into multiple columns

#font_import()

mytheme <- theme(axis.line = element_line(colour = "black"),
                 panel.grid.major = element_blank(),
                 panel.grid.minor = element_blank(),
                 panel.border = element_blank(),
                 panel.background = element_blank(),
                 text=element_text(size=20,  family="ArialMT", colour = "black"),
                 axis.title.x = element_text(vjust=-0.5),
                 axis.title.y = element_text(vjust=0.5), 
                 legend.position = "bottom")

pie_theme <- theme_minimal()+
  theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.border = element_blank(),
  panel.grid=element_blank(),
  axis.ticks = element_blank(),
  plot.title=element_text(size=14, face="bold")
  )
```

```{r}
Targeted_Seq_1_wkd <- "/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/Targeted/Targeted_Seq_1"
Individual <- "/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/WholeTranscriptome/Individual/Isoseq3.2.1"

Probes_Match <- read.table(paste0(Targeted_Seq_1_wkd,"/MAPPING/K18.fasta.sam.probe_hit.txt"), 
                  header=T, as.is=T, sep="\t")
```


```{r}
colnames(Probes_Match)[6] <- "Probes"
Probes_Match$Gene <- word(Probes_Match$Probes,c(3),  sep = fixed ('_'))
Probes_Match$Gene <- word(Probes_Match$Gene,c(1),  sep = fixed ('('))

p3 <- Probes_Match %>%
  filter(num_probe > 0) %>% 
  group_by(Gene) %>% 
  tally() %>% 
  ggplot(., aes(x = reorder(Gene, -n), y = n)) + 
  geom_bar(stat = "identity") + 
  theme_bw() + 
  mytheme + 
  geom_text(aes(label=n), position=position_dodge(width=0.9), vjust=-0.25) +
  labs(x = "", y = "Number of Transcripts", title = "Number of Transcripts per Gene identified \n\n\n" ) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

p3
```


```{r}
Tofu_ignored <- read.table(paste0(Targeted_Seq_1_wkd,"/TOFU/K18.ignored_ids.txt"), 
                  header=F, as.is=T, sep="\t")
colnames(Tofu_ignored) <- c("read_id", "Rationale")

##### multiple transcripts from Polished!
#n_occur <- data.frame(table(Tofu_ignored$read_id))
Tofu_ignored <- aggregate(Rationale ~ read_id, data = Tofu_ignored, paste, collapse = ",")

n_occur <- data.frame(table(Probes_Match$read_id))
#n_occur[n_occur$Freq > 1,]
#Probes_Match[Probes_Match$read_id == "transcript/954", ]
#Probes_Match[Probes_Match$read_id == "transcript/1993", ]

Probes_Match <- merge(Probes_Match,Tofu_ignored, by = "read_id", all.x = T)
Multiple_hits <- data.frame()
#for(i in n_occur[n_occur$Freq > 1,1]){
  #Multiple_hits <- rbind(Multiple_hits, Probes_Match[Probes_Match$read_id == i,])
```


```{r}
p4 <- Probes_Match %>%
  filter(!is.na(Rationale)) %>%
  ggplot(.,aes(x = Gene)) + 
  geom_bar() + 
  theme_bw() + 
  mytheme + 
  labs(x = "", y = "Number of Transcripts Removed", title = "Number of Transcripts Removed after Tofu Collapse \n\n\n" )

p4
  
Probes_Match$Retained <- ifelse(is.na(Probes_Match$Rationale), "Yes", "No")
```


```{r}
p6 <- Probes_Match %>%
  filter(num_probe > 0) %>% 
  group_by(Gene, Retained) %>% 
  tally() %>% 
  ggplot(., aes(x = reorder(Gene, -n), y = n, fill = Retained)) + 
  geom_bar(stat = "identity") + 
  theme_bw() + 
  mytheme + 
  labs(x = "", y = "Number of Transcripts", title = "Number of Transcripts per Gene identified \n\n\n" ) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

p6
```
```{r}
Tofu_collapsed <- read.table(paste0(Targeted_Seq_1_wkd,"/TOFU/K18.collapsed.group.txt"), 
                  header=F, as.is=T, sep="\t")
colnames(Tofu_collapsed) <- c("pb.id", "read_id")

# long to wide 
Tofu_collapsed <- cSplit(Tofu_collapsed,"read_id", ",")
## note warning for gather as different lengths of collapse 
Tofu_collapsed <- gather(Tofu_collapsed, remove, read_id, read_id_01: colnames(Tofu_collapsed)[dim(Tofu_collapsed)[2]], factor_key=TRUE)
Tofu_collapsed <- Tofu_collapsed[!is.na(Tofu_collapsed$read_id),]
Tofu_collapsed <- Tofu_collapsed[,-2]

Probes_Match <- merge(Probes_Match,Tofu_collapsed, by = "read_id", all = T)
head(Probes_Match)
View(Probes_Match %>% count(pb.id))
```
```{r}
target_pre_sqanti <- paste0(Targeted_Seq_1_wkd,"/SQANTI2/K18.collapsed.filtered.rep_classification.txt") 
whole_sqanti <- paste0(Individual, "/SQANTI2/K18.collapsed.filtered.rep_classification.txt")
target_sqanti <- paste0(Targeted_Seq_1_wkd,"/SQANTI2/TargetedSeq1.collapsed.filtered.rep_classification.filtered_lite_classification.txt")
source("/gpfs/ts0/home/sl693/general/QC_Figures/Rscripts/SQANTI_General.R")
#class.file <- SQANTI_class_preparation(sqanti)
target_pre.class.file <- SQANTI_class_preparation(target_pre_sqanti)
whole_pre.class.file <- SQANTI_class_preparation(whole_sqanti)

#class.file$associated_gene <- toupper(class.file$associated_gene)
target_pre.class.file$associated_gene <- toupper(target_pre.class.file$associated_gene)
whole_pre.class.file$associated_gene <- toupper(whole_pre.class.file$associated_gene)

targeted_genes <- c("APP", "CLU", "APOE", "MAPT", "BIN1", "SORL1", "TARDBP", 
                    "PICALM", "PTK2B", "FYN","SNCA", "FUS", "ABCA1",
                    "VGF", "ABCA7", "TREM2", "ANK1", "CD33", "RHBDF2", "MAPT")
# identifiers 
target_pre.class.file$ID <- paste(target_pre.class.file$associated_gene,
                                  target_pre.class.file$associated_transcript,target_pre.class.file$structural_category,
                                     target_pre.class.file$CDS_genomic_start,target_pre.class.file$CDS_genomic_end, sep = "_")
whole_pre.class.file$ID <- paste(whole_pre.class.file$associated_gene,
                                 whole_pre.class.file$associated_transcript,whole_pre.class.file$structural_category,
                                   whole_pre.class.file$CDS_genomic_start,whole_pre.class.file$CDS_genomic_end, sep = "_")

transcript_expression <- merge(whole_pre.class.file[,c("Log_ISOSEQ_TPM","ID")], target_pre.class.file[,c("Log_ISOSEQ_TPM","ID")], by = "ID", all = T)
colnames(transcript_expression) <- c("ID","Whole_Expression","Targeted_Expression")

for(i in 1:nrow(transcript_expression)){
  transcript_expression$Detection[i] <- 
  if(!is.na(transcript_expression$Targeted_Expression[i]) & !is.na(transcript_expression$Whole_Expression[i])){
    paste("Both")
  }else if (!is.na(transcript_expression$Targeted_Expression[i]) & is.na(transcript_expression$Whole_Expression[i])){
      paste("Targeted")
  }else{
      paste("Not_Targeted")
  }
}

#transcript_expression <- gather(transcript_expression, Approach, Expression_Log_ISOSEQ_TPM, Whole_Expression:Targeted_Expression, factor_key=TRUE)
datatable(transcript_expression)

# replace Na values for plot 
transcript_expression$Whole_Expression[is.na(transcript_expression$Whole_Expression)] <- 0
transcript_expression$Targeted_Expression[is.na(transcript_expression$Targeted_Expression)] <- 0

ggplot(transcript_expression, aes(Whole_Expression, Targeted_Expression, colour = Detection)) +
  geom_point() + 
  mytheme + 
  labs(title = "IsoSeq Transcript Abundance \nfrom Targeted and Whole Transcriptome Approach\nTranscript Abundance (FL in TPM)\n\n\n", 
       y = "Targeted", x = "Whole" )



## Boxplot 
ggplot(transcript_expression, aes(x = Detection, y = Whole_Expression)) + 
  geom_boxplot() +
  mytheme + 
  labs(title = "Transcript abundance in whole transcriptome\n\n", x = "", y = "Transcript Abundance (FL in TPM)") +
  scale_x_discrete(labels=c("Detected in Targeted","Not Detected in Targeted","Only Targeted"))


# Length 
transcript_length <- merge(whole_pre.class.file[,c("length","ID")], target_pre.class.file[,c("length","ID")], by = "ID", all = T)
colnames(transcript_length) <- c("ID","Whole_Length","Targeted_Length")
for(i in 1:nrow(transcript_length)){
  transcript_length$Detection[i] <- 
  if(!is.na(transcript_length$Targeted_Length[i]) & !is.na(transcript_length$Whole_Length[i])){
    paste("Both")
  }else if (!is.na(transcript_length$Targeted_Length[i]) & is.na(transcript_length$Whole_Length[i])){
      paste("Targeted")
  }else{
      paste("Not_Targeted")
  }
}
ggplot(transcript_length, aes(x = Detection, y = Whole_Length)) + 
  geom_boxplot() +
  mytheme + 
  labs(title = "Transcript Length in whole transcriptome\n\n", x = "", y = "Transcript Length (bases)") +
  scale_x_discrete(labels=c("Detected in Targeted","Not Detected in Targeted","Only Targeted"))

```


```{r}
off_target_pre.class.file <- target_pre.class.file[!target_pre.class.file$associated_gene %in% targeted_genes,]
# number of off-target transcripts
nrow(off_target_pre.class.file)
# number of off-target genes 
nrow(off_target_pre.class.file %>% count(associated_gene))

# chromosomal distribution of off-targets at transcript level to ensure random 
off_target_pre.class.file %>%
  count(chrom) %>%
  ggplot(., aes(x = chrom, y = n)) + 
  geom_bar(stat = "identity") + 
  mytheme + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  scale_x_discrete(limits=c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14",
                            "chr15","chr16","chr17","chr18","chr19","chrM","chrX")) + 
  labs(y = "Frequency of Transcripts", x = "", title = "Chromosomal distribution of off-target genes\n\n\n")

# chromosomal distribution of off-targets at gene level to ensure random 
off_target_genechrom <- off_target_pre.class.file %>%
  group_by(associated_gene) %>%
  count(chrom) 

off_target_genechrom  %>% 
  ggplot(., aes(x = chrom, y = n)) + 
  geom_bar(stat = "identity") + 
  mytheme + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  scale_x_discrete(limits=c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14",
                            "chr15","chr16","chr17","chr18","chr19","chrM","chrX")) + 
  labs(y = "Frequency of Genes", x = "", title = "Chromosomal distribution of off-target genes\n\n\n")


off_target_genechrom %>%
  group_by(chrom) %>%
  tally() %>%
  mutate(perc = n/sum(n) * 100) %>%
  ggplot(., aes(x = chrom, y = perc)) + 
  geom_bar(stat = "identity") + 
  mytheme + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  scale_x_discrete(limits=c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14",
                            "chr15","chr16","chr17","chr18","chr19","chrM","chrX")) + 
  labs(y = "Percentage of Genes", x = "", title = "Chromosomal distribution of off-target genes\n\n\n")

```
