---
title: "DGE/DTU - IsoSeq Raw Data"
author: Szi Kay Leung
date: 'Report created: `r Sys.Date()`'
output:
  rmdformats::material:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: true
    highlight: tango
    code_folding: hide
    fig_caption: true
    cards: true
---

```{r preambleImports, echo=FALSE, results='hide', warning=FALSE, message=FALSE}
# This R code is based on R (and python) methods distributed with 
#       https://github.com/nanoporetech/pipeline-transcriptome-de
#
# General review and evaluation of the starting sequences and the
# experimental design has been copied from the DESeq2 tutorial in
# this same series

knitr::opts_chunk$set(echo = TRUE,warning=FALSE,message=FALSE,fig.width=14, fig.height= 7)
setwd("/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/WholeTranscriptome/Individual/Isoseq/Isoseq3_WKD")

library(yaml)
library(DRIMSeq)
library(AnnotationDbi)
library(GenomicFeatures)
library(dplyr)
library(edgeR)
library(DEXSeq)
library(devtools)
library(ggplot2)
library(tidyr)
library(session)     # session persistence added as feature request
library(writexl)
library(digest)
library(kableExtra)
library(stageR)
library(ShortRead)
library(stringr)
# BiocInstaller::biocLite(pkgs=c("tidyR"))
```

```{r fileimport, echo=FALSE, results='hide', warning=FALSE, message=FALSE}
config <- yaml.load_file("/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/Scripts/IsoSeq3_Tg4510/Differential_Analysis/config_isoseq_raw.yaml")
resultDir <- paste0(config$output_dir,"/Results")


```

# Aim

This analysis has been adapted from **`Nanopore_Transcriptome_Tutorial.Rmd`*:  
- Identify the gene isoforms that are expressed and the number of sequence reads that are observed from each gene isoform (and parental gene).  
- Map sequence reads to the reference transcriptome using Minimap2 and quantify the number of reads associated with each isoform/gene using Salmon
- Perform a statistical analysis using edgeR (McCarthy et al. (2012)) to identify differentially expressed genes and DEXSeq (Reyes et al. (2013)) and stageR (Van den Berge and Clement (2018)) to identify differentially used transcripts.

# Study design

The study design described by the configuration file is presented in the table below.


```{r setupParameters, echo=FALSE}
studyDesign <- data.frame()
for (i in 1:length(config$Samples)) {
  studyDesign <- rbind(studyDesign, 
                       data.frame(samples=names(config$Samples[[i]][[1]]), 
                                  filename=unlist(config$Samples[[i]][[1]]), 
                                  group=names(config$Samples[[i]])))
}

studyDesign$replicate <- sapply(1:nrow(studyDesign), function(x)sum(studyDesign$group[1:x]==studyDesign$group[x]))
# let's use the provided filename as the key??
studyDesign$group <- relevel(studyDesign$group, ref=config$referenceGroup)
# quick tidy
studyDesign <- studyDesign[,-which(colnames(studyDesign)=="samples")]

knitr::kable(studyDesign, booktabs=TRUE, table.envir='table*', linesep="")  %>%
  kable_styling(latex_options=c("hold_position", "scale_down"))

```

*Table showing the cDNA sequence files evaluated for DGE and DTU. Treatment group and replicate number are shown for each of the sequence files.*

# Review of length and mapping 

## {.tabset .tabset-pills}

### Raw cDNA sequences 

The summary statistics displayed include observational metrics such as the number of sequence reads, information on the longest, shortest and mean read lengths, GC content, N50 and N90 read lengths. Ideally sequences from each library will have similar characteristics. Large differences in e.g. the number of reads, overall read quality and GC content may indicate technical differences between the samples - such differences may confound the statistical analysis and hinder the identification of differentially expressed genes and transcripts.


```{r seqLibraryReview, echo=FALSE}
#The observations shown in the table below have been produced using the R ShortRead package (@R-ShortRead). 
ncalc <- function(len.vector, n) {
  # N50 - length such that scaffolds of this length or longer include half the bases of the assembly
  len.sorted <- rev(sort(len.vector))
  len.sorted[cumsum(len.sorted) >= sum(len.sorted)*n][1]
}

lcalc <- function(len.vector, n) {
  len.sorted <- rev(sort(len.vector))
  which(cumsum(len.sorted) >= sum(len.sorted)*n)[1]
}

processQCFastq <- function(rowname) {
  row <- which(row.names(studyDesign)==rowname)
  file <- as.character(studyDesign[row, "filename"])
  fastq <- readFastq(paste0(config$input_dir,file))
  c(
    reads = formatC(length(fastq), big.mark=","),
    mbs = formatC(round(sum(width(fastq)) / 1000 / 1000, digits=1), big.mark=","),
    min = min(width(fastq)),
    max = max(width(fastq)),
    mean = round(mean(width(fastq)), digits=1),
    median = round(median(width(fastq)), digits=0),
    qval = round(mean(alphabetScore(fastq) / width(fastq)), digits=1),
    gc = round(mean(letterFrequency(sread(fastq), "GC")  / width(fastq)) * 100, digits=1),
    n50 = ncalc(width(fastq), n=0.5),
    l50 = lcalc(width(fastq), n=0.5),
    n90 = ncalc(width(fastq), n=0.9),
    l90 = lcalc(width(fastq), n=0.9)
  )
}

data <- lapply(row.names(studyDesign), processQCFastq)
qcData <- data.frame(data)
colnames(qcData) <- row.names(studyDesign)

knitr::kable(qcData, booktabs=TRUE, table.envir='table*', linesep="")  %>%
  kable_styling(latex_options=c("hold_position", font_size=9)) 
```

*Table presenting summary statistics for the cDNA libraries imported*


### Review of cDNA read mapping

These mapping statistics are useful for evaluating the concordance of different sequence libraries and thus the overall quality of the study - ideally there should be robust similarities in the fraction of mapping sequence reads / multi-mapping reads etc.

The read mapping statistics shown in the table above are derived from the **`bam`** files produced by **`minimap2`**. **`read mappings`** corresponds to the number of unique read mappings described in the BAM file (one read may map to more than one location). **`Secondary`** corresponds to the number of secondary alignment for a read that can be mapped to multiple locations. A **`Supplementary`** alignment could correspond to an alignment that is chimeric or represents a structural variant or complex fusion event. A **`Duplicate`** read is unlikely but would represent multiple *identical* reads at the same locus that are likely the result of an amplification during the library preparation. **`%Mapping`** corresponds to the percentage fraction of sequence reads that are mapped to the reference genome - this is calculated as the number of primary mappings against the total number of sequence reads observed in the starting fastq files.

```{r flagstatStats, echo=FALSE, cache=FALSE}
#The workflow includes a **`samtools flagstat`** mapping statistic summary, written to files in the `Analysis/samtools` folder.
flagstatTargets <- file.path(config$output_dir, "/samtools", 
    paste(tools::file_path_sans_ext(word(basename(as.character(studyDesign$filename)),c(1),sep=fixed(".")), compression=TRUE),".flagstat",sep="")
)
loadFlagstat <- function(file) {
  x <- read.table(file, header=FALSE, sep=" ", fill=NA)[c(1:5),c(1,3)]
  x[,1]
}

flagstatRes <- data.frame(matrix(unlist(lapply(flagstatTargets, loadFlagstat)), ncol=length(flagstatTargets)), stringsAsFactors = FALSE)
colnames(flagstatRes) <- rownames(studyDesign)
rownames(flagstatRes) <- c("read mappings", "Secondary", "Supplementary", "Duplicates", "Mapped")

# include the measured number of reads; this is real
flagstatRes[nrow(flagstatRes)+1,] <- as.numeric(gsub(",","",t(qcData)[, "reads"]))
rownames(flagstatRes)[6] <- "nreads"

getVal <- function(word) {
  sum(as.numeric(unlist(strsplit(word, "/"))))
}

zreads <- unlist(lapply(flagstatRes["read mappings", ], getVal)) -
  unlist(lapply(flagstatRes["Secondary", ], getVal)) -
  unlist(lapply(flagstatRes["Supplementary", ], getVal)) - 
  unlist(lapply(flagstatRes["Duplicates", ], getVal)) 

flagstatRes[nrow(flagstatRes)+1,] <- zreads
rownames(flagstatRes)[7] <- "zreads"

flagstatRes[nrow(flagstatRes)+1,] <- as.numeric(flagstatRes["zreads", ]) / as.numeric(flagstatRes["nreads", ]) * 100
rownames(flagstatRes)[8] <- "%mapping"

flagstatRes <- flagstatRes[c(6,1,2,3,4,8),]

knitr::kable(flagstatRes, booktabs=TRUE, table.envir='table*', linesep="")  %>%
  kable_styling(latex_options=c("hold_position", font_size=11))

```

*Table showing summary statistics from the minimap2 long read mapping of reads to the reference transcriptome*






```{r, echo=FALSE}
## this code block is intended to replace the merge_counts step of the Snakefile
## Python code in Snakefile splits the name on the 1st "."
samples <- gsub("\\..+", "", basename(as.character(studyDesign$filename)))
## this code block to replace the col_data export and import
coldata <- data.frame(cbind(NULL, sample_id=samples, condition=as.character(studyDesign$group), type="single-read"), stringsAsFactors=TRUE)


```

# Salmon: Assigning mapped cDNA sequence reads to transcripts

cDNA sequence reads have been mapped to the reference transcriptome using **`minimap2`**. The **`Salmon`** tool (@Salmon2017) has been used to assign cDNA read counts to the annotated transcripts defined within the transcriptome **`fasta`** file provided. This was performed using the **`alignment`**-based mode of Salmon. As shown, Salmon has assigned expressed cDNA sequence reads to individual transcripts or gene isoforms. 

```{r importSalmonCounts, results='hide', echo=FALSE}
# import list of path directory containining individual output files from salmon
salmonCountFiles <- paste(file.path(config$output_dir,"/Salmon",samples,"quant.sf"))

rawSalmonCounts <- data.frame(Reference=character(), stringsAsFactors = FALSE)
harvestSalmonCounts <- function(file) {
  # read in individual salmon file, only keeping reads with more than 0 counts 
  obs <- read.table(file, header=TRUE, stringsAsFactors = FALSE)
  # merge to existing dataframe 
  rawSalmonCounts <<- full_join(rawSalmonCounts, obs[which(obs$NumReads>0), c("Name", "NumReads")], by=c("Reference"="Name"))
  file
}
lapply(salmonCountFiles, harvestSalmonCounts)
colnames(rawSalmonCounts) <- append("Reference", samples) # replace merging column names with sample names
rownames(rawSalmonCounts) <- gsub("\\..+", "", as.character(rawSalmonCounts$Reference)) # rownames for excel export

# adjust column names for the presented table and excel file
vRawSalmonCounts <- rawSalmonCounts[order(rowSums(rawSalmonCounts[, seq(2, ncol(rawSalmonCounts))]), decreasing=TRUE),]
newnames <- rownames(studyDesign)[match(colnames(vRawSalmonCounts)[which(colnames(vRawSalmonCounts) %in% samples)], samples)]
colnames(vRawSalmonCounts)[which(colnames(vRawSalmonCounts) %in% samples)] <- newnames

knitr::kable(vRawSalmonCounts[1:10,], booktabs=TRUE, table.envir='table*', linesep="") %>%
  kable_styling(latex_options=c("hold_position", font_size=11))

csvExpressedTranscripts <- file.path(resultDir, "ExpressedTranscripts.csv")
#write.csv(vRawSalmonCounts, file = csvExpressedTranscripts)

```

*Table showing the 10 annotated transcript features identified by Minimap2 mapping and Salmon transcript detection with the highest number of mapped reads*

## Isoforms to Genes
The next step in the sequence analysis is to map gene isoforms back to their parental genes for gene level analysis. This utilises the genome annotation **`gff`** file described in the **`config.yaml`** file.


```{r loadAnnotations, echo=FALSE, results='hide', warning=FALSE, message=FALSE}
# load the database annotation from provided GFF file
txdb <- makeTxDbFromGFF(config$annotation)  # read Annotation field from config.yaml
txdf <- AnnotationDbi::select(txdb, keys(txdb,"GENEID"), "TXNAME", "GENEID")
tab <- table(txdf$GENEID)
txdf$ntx<- tab[match(txdf$GENEID, names(tab))]

# FIXME: filter for transcripts which are in the annotation. Why they are not all there? 
filteredSalmonCounts <- rawSalmonCounts[rawSalmonCounts$Reference %in% txdf$TXNAME == TRUE,]

# Create counts data frame:
salmonCounts<-data.frame(txdf[match(filteredSalmonCounts$Reference, txdf$TXNAME), c(1,2)], filteredSalmonCounts)
salmonCounts[is.na(salmonCounts)] <- 0
colnames(salmonCounts) <- append(c("gene_id", "feature_id"),colnames(filteredSalmonCounts))

# adjust column names for the presented table and excel file
vSalmonCounts <- salmonCounts[order(rowSums(salmonCounts[, seq(4, ncol(salmonCounts))]), decreasing=TRUE),]
newnames <- rownames(studyDesign)[match(colnames(vSalmonCounts)[which(colnames(vSalmonCounts) %in% samples)], samples)]
colnames(vSalmonCounts)[which(colnames(vSalmonCounts) %in% samples)] <- newnames

csvExpressedGenesTranscripts <- file.path(resultDir, "ExpressedGenesTranscripts.csv")
#write.csv(vSalmonCounts, file = csvExpressedGenesTranscripts)

```

This step is important for the statistical analyses and also reveals the number of transcripts that can be observed for each gene. While a majority of genes contain a single expressed isoform (**`r round(as.numeric((table(table(salmonCounts$gene_id)) / length(table(salmonCounts$gene_id)) * 100)[1]), digits=1)` %** of genes contain a single isoform). The largest number of transcripts associated with a single gene is **`r max(table(salmonCounts$gene_id))` ** transcripts.

```{r geneTranscriptCount, echo=FALSE}
plotData(dmDSdata(counts=salmonCounts, samples=coldata))
```

*Frequency plot showing the number of transcripts assigned to genes across the sampled cDNA sequence collection*

# DRIMSeq: Read filtering

**`DRIMSeq`** (@R-DRIMSeq) is used to filter the transcript count data from the **`salmon`** analysis. The filter step will be used to select for genes and transcripts that satisfy rules for the number of samples in which a gene or transcript must be observed and minimum threshold levels for the number of observed reads. The parameters defined for this analysis include

* `min_samps_gene_expr` = **`r config$min_samps_gene_expr`** - a transcript must be mapped to a *gene* in at least this *minimum number of samples* for the *gene* be included in the analysis
* `min_samps_feature_expr` = **`r config$min_samps_feature_expr`** - a transcript must be mapped to an *isoform* in at least this this *minimum number of samples* for the *gene isoform* to be included in the analysis
* `min_gene_expr` = **`r config$min_gene_expr`** - the *minimum number of total mapped sequence reads* for a gene to be considered *expressed*
* `min_feature_expr` = **`r config$min_feature_expr`** - the *minimum number of total mapped sequence reads* for a *gene isoform* to be considered

```{r, echo=FALSE, warning=FALSE}
dmDSdataObj  <- dmDSdata(counts=salmonCounts, samples=coldata)
trs_cts_unfiltered <- counts(dmDSdataObj)

dmDSdataObj <- dmFilter(dmDSdataObj, min_samps_gene_expr = config$min_samps_gene_expr, 
              min_samps_feature_expr = config$min_samps_feature_expr,
              min_gene_expr = config$min_gene_expr, 
              min_feature_expr = config$min_feature_expr)



# Sum transcript counts into gene counts:
trs_cts <- counts(dmDSdataObj)
names(trs_cts)[2]<-"transcript_id"  # changed from featureID

gene_cts <- trs_cts_unfiltered %>% dplyr::select(c(1, 3:ncol(trs_cts)))  %>% group_by(gene_id) %>% summarise_all(list(sum)) %>% data.frame()
rownames(gene_cts) <- gene_cts$gene_id
gene_cts$gene_id <- NULL

# write out these feature level (isoform) and gene count data to Excel files
xlsExpressedGeneCounts <- file.path(resultDir, "FilteredExpressedGeneCounts.xlsx")
#write_xlsx(gene_cts, path = xlsExpressedGeneCounts)

xlsExpressedFeatureCounts <- file.path(resultDir, "FilteredExpressedFeatureCounts.xlsx")
#write_xlsx(trs_cts, path = xlsExpressedFeatureCounts)

```


# EdgeR: Differential gene expression analysis

A statistical analysis is first performed using **`edgeR`** (@R-edgeR2010, @R-edgeR2012) to identify the subset of differentially expressed genes. The filtered list of gene counts is used as input. A normalisation factor is calculated for each sequence library (using the default `TMM` method - please see @R-edgeR2012 for further details). The defined experimental design is used to calculate estimates of dispersion across for each of the gene features. Genewise statistical tests are then calculated, again using the contrasts defined in the experimental design. The differentially expressed genes are corrected for false discovery (*fdr*) using the method of Benjamini & Hochberg (@BH1995) 


```{r, echo=FALSE, results='hide', warning=FALSE, error=FALSE}
# Building model matrix - ensuring that the config.yaml specified reference is key in specified factor ...
Group <- factor(studyDesign$group, 
                levels=append(config$referenceGroup,
                              levels(studyDesign$group)[which(levels(studyDesign$group)!=config$referenceGroup)]))
design <- model.matrix(~Group)
# Differential gene expression using edgeR:

y <- DGEList(gene_cts)
y <- calcNormFactors(y)
y <- estimateDisp(y,design)
fit <- glmQLFit(y,design)

qlf <- glmQLFTest(fit)
edger_res <- topTags(qlf, n=nrow(y), sort.by="PValue")[[1]]

# Convert ENMUSG to gene name 
reference <- read.csv("/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/reference_2019/gencode.vM22_gene_annotation_table.csv") %>%
  mutate(Gene = word(GeneSymbol, c(3), sep = fixed(" "))) %>% 
  mutate(Transcript = word(Transcriptid, c(3), sep = fixed(" ")))
knitr::kable(edger_res[1:15,], digits = c(2,2,2,45,45), booktabs=TRUE, table.envir='table*', linesep="")  %>%
  kable_styling(latex_options=c("hold_position"), font_size=9)

# Export output
xlsEdgeRDiffExpr <- file.path(resultDir, "EdgeRDiffExpression.xlsx")
#write_xlsx(as.data.frame(edger_res), path = xlsEdgeRDiffExpr)
```

*Table showing the top 15 genes, ranked by adjusted p-value, from the edgeR analysis. Information shown includes the log2 fold change between experimental conditions, the log-scaled counts per million measure of abundance and the false discovery corrected p-value (FDR). This table has not been filtered for genes that satisfy statistical or magnitudinal thresholds*


The results from the **`edgeR`** analysis can be plotted as an **`MA Plot`**. This plot visualises differences in measurements between the two experimental conditions. M is the log2 ratio of gene expression calculated between the conditions. A is a log2 transformed mean expression value. The figure below presents the MA figure from this edgeR analysis. Genes that satisfy the *logFC* and FDR corrected *p-value* thresholds defined in the **`config.yaml`** configuration file are shaded. 


```{r MAGenePlot, echo=FALSE}

MAPlotData <- qlf$table
logUp <- which(MAPlotData$logFC >= config$lfcThreshold)
logDown <- which(MAPlotData$logFC <= -config$lfcThreshold)
withStat <- which(MAPlotData$PValue <= config$adjPValueThreshold)
colours <- c(noDifference="gray", upRegulated="red", downRegulated="green")
gene <- rep("noDifference", nrow(MAPlotData))
gene[logUp[logUp %in% withStat]] <- "upRegulated"
gene[logDown[logDown %in% withStat]] <- "downRegulated"

plot <- ggplot(data.frame(MAPlotData), aes(y=logFC, x=logCPM)) + 
    geom_point(size=1.2) + 
    geom_hline(yintercept=-config$lfcThreshold, color="green") + 
    geom_hline(yintercept=config$lfcThreshold, color="red") + 
    aes(colour=gene) + 
    scale_colour_manual(values=colours) +
    ggtitle("MA plot showing logFC against logCPM for gene level data")

print(plot)
#plotQLDisp(fit)
```


# DeSeq2: Differential gene expression analysis
```{r}
#coldata <- data.frame("group" = studyDesign$group)
#rownames(coldata) <- colnames(gene_cts)
#dds <- DESeqDataSetFromMatrix(countData = as.matrix(gene_cts),
                              #colData = coldata,
                              #design= ~ group)

```

# Differential transcript usage analysis

Differential transcript usage analysis is performed using the *R* **`DEXSeq`** package (@R-DEXSeq). Similar to the **`edgeR`** package, **`DEXSeq`** estimates the variance between the biological replicates and applies generalised linear models for the statistical testing. The key difference is that the **`DEXSeq`** method is looking for differences at the exon count level.

**`DEXSeq`** uses the filtered transcript count data prepared earlier in this analysis. 

```{r DEXSeq, echo=FALSE, warning=FALSE, error=FALSE, results='hide', message=FALSE}

# Differential transcript usage using DEXSeq:

sample.data<-DRIMSeq::samples(dmDSdataObj)
count.data <- round(as.matrix(counts(dmDSdataObj)[,-c(1:2)]))
dxd <- DEXSeqDataSet(countData=count.data, sampleData=sample.data, design=~sample + exon + condition:exon, featureID=trs_cts$transcript_id, groupID=trs_cts$gene_id)

dxd <- estimateSizeFactors(dxd)
dxd <- estimateDispersions(dxd)
dxd <- testForDEU(dxd, reducedModel=~sample + exon)
dxd <- estimateExonFoldChanges( dxd, fitExpToVar="condition")

dxr <- DEXSeqResults(dxd, independentFiltering=FALSE)
colnames(dxr)[grep("log2fold", colnames(dxr))] <- "log2fold"

```



```{r diffIsofromExprTable, echo=FALSE}
MADTUdata <- data.frame(dxr)[order(dxr$padj),c("exonBaseMean", "log2fold", "pvalue", "padj")]
MADTUdata$exonBaseMean <- log2(MADTUdata$exonBaseMean)
colnames(MADTUdata)[which(colnames(MADTUdata)=="exonBaseMean")] <- "Log2MeanExon"
colnames(MADTUdata)[which(colnames(MADTUdata)=="log2fold")] <- "Log2FC"

knitr::kable(MADTUdata[1:15,], digits = c(2,2,45,45), booktabs=TRUE, table.envir='table*', linesep="")  %>%
  kable_styling(latex_options=c("hold_position"), font_size=9)
```

*Table showing the top 15 gene isoforms, ranked by adjusted p-value, from the DEXSeq analysis. Information shown includes the log2 fold change between experimental conditions, the log-scaled transcript abundance and the false discovery corrected p-value (FDR). This table has not been filtered for genes that satisfy statistical or magnitudinal thresholds*



```{r MAPlotIsoform, echo=FALSE}

#plotMA(dxr, cex=0.8, alpha=0.05) 


logUp <- which(MADTUdata$Log2FC >= config$lfcThreshold)
logDown <- which(MADTUdata$Log2FC <= -config$lfcThreshold)
withStat <- which(MADTUdata$padj <= config$adjPValueThreshold)
colours <- c(noDifference="gray", upRegulated="red", downRegulated="green")
gene <- rep("noDifference", nrow(MADTUdata))
gene[logUp[logUp %in% withStat]] <- "upRegulated"
gene[logDown[logDown %in% withStat]] <- "downRegulated"

plot <- ggplot(MADTUdata, aes(y=Log2FC, x=Log2MeanExon)) + 
    geom_point(size=1.2) + 
    geom_hline(yintercept=-config$lfcThreshold, color="green") + 
    geom_hline(yintercept=config$lfcThreshold, color="red") + 
    aes(colour=gene) + 
    scale_colour_manual(values=colours) +
    ggtitle("MA plot showing logFC against log count for isoform level data")+ 
    ylab("M (log2 transformed differential expression)")  + 
    xlab("A (log2 transformed mean exon read counts)")

print(plot)

```

<!--
deprecated the display of the dispersal estimates - not for an introductory tutorial...
plotDispEsts(dxd, cex=0.8)
-->


```{r xlsAlternativelySplicedGenes, echo=FALSE}
xlsAlternativelySplicedGenes <- file.path(resultDir, "AlternativelySplicedGenes.xlsx")
alternativelySplicedGenes <- data.frame(gene_id=names(perGeneQValue(dxr)) , perGeneQValue(dxr), row.names=NULL)[order(perGeneQValue(dxr)), ]
#write_xlsx(alternativelySplicedGenes, path = xlsAlternativelySplicedGenes)

xlsDiffExpressedIsoforms <- file.path(resultDir, "DiffExpressedIsoforms.xlsx")
diffExpressedIsoforms <- dxr[order(dxr$padj), c("featureID", "groupID", "padj")]
#write_xlsx(data.frame(diffExpressedIsoforms), path = xlsDiffExpressedIsoforms)
```

~~~
`r xlsAlternativelySplicedGenes`
~~~

The Excel format spreadsheet file indicated above contains the output of the **`DEXSeq`** DTU analysis.


# stageR analysis

The final component of this cDNA sequence based gene isoform analysis is a stage-wise statistical analysis using the **`R`** software package called **`stageR`** (@R-stageR). **`stageR`** uses (1) the raw p-values for DTU from the **`DEXSeq`** analysis in the previous section and (2) a false-discovery corrected set of p-values from testing whether individual genes contain at least one exoning showing DTU. A hierarchical two-stage statistical testing evaluates the set of genes for DTU.    



```{r stageRanalysis, echo=FALSE, warning=FALSE, message=FALSE}
# stageR analysis of DEXSeq results:
pConfirmation <- matrix(dxr$pvalue, ncol=1)
dimnames(pConfirmation) <- list(dxr$featureID, "transcript")
pScreen <- perGeneQValue(dxr)
tx2gene <- as.data.frame(dxr[,c("featureID", "groupID")])

stageRObj <- stageRTx(pScreen=pScreen, pConfirmation=pConfirmation, pScreenAdjusted=TRUE, tx2gene=tx2gene)
# note: the choice of 0.05 here means you can *only* threshold at 5% OFDR later
stageRObj <- stageWiseAdjustment(stageRObj, method="dtu", alpha=0.05)
dex.padj <- getAdjustedPValues(stageRObj, order=FALSE, onlySignificantGenes=FALSE)
colnames(dex.padj) <- c("gene_id", "transcript_id", "p_gene", "p_transcript")

xlsStageRDiffTranscriptUsage <- file.path(resultDir, "StageRDiffTranscriptUsage.xlsx")
#write_xlsx(dex.padj, path = xlsStageRDiffTranscriptUsage)
```

~~~
`r xlsStageRDiffTranscriptUsage`
~~~

The matrix of gene and transcript identifiers and the probability that they are showing any form of DTU has been written to an Excel format spreadsheet file at the path shown above.


```{r stageRSummary, echo=FALSE}
candidates <- union(which(dex.padj$p_gene < config$adjPValueThreshold), which(dex.padj$p_transcript < config$adjPValueThreshold))
candidateList <- dex.padj[candidates,]

# filter out transcripts that are not showing DTU - will lose a couple of genes in the process ...
candidateList <- candidateList[which(candidateList$p_transcript <= config$adjPValueThreshold), ]

candidateList <- candidateList[order(candidateList$p_transcript),]

knitr::kable(candidateList, digits = c(2,2,5,5), booktabs=TRUE, table.envir='table*', linesep="")  %>%
  kable_styling(latex_options=c("hold_position"), font_size=10)
```

*Table showing gene and transcript identifiers and their FDR corrected probabilities for the genes and their isoforms that have been identified as showing DTU using the R packages DEXSeq and StageR. This list has been shortened requiring that both gene and transcript must satisfy the p-value threshold*


The characteristics of the genes showing DTU can be summarised graphically. The following subsections show example genes (selected arbitrarily) that show DGE and DTU and the expression patterns for their annotated transcripts. 


```{r mungStageR, echo=FALSE, warning=FALSE, message=FALSE, results='asis'}
# Join counts and stageR results:
df <- trs_cts %>% left_join(dex.padj, by = c("gene_id", "transcript_id"))
df <- df[order(df$p_gene),]

# quickie to pull out the sample ids (could be derived from the earlier annotations?)
scols <- setdiff(names(df),c("gene_id", "transcript_id", "p_gene", "p_transcript"))
# Normalise counts???
ndf <- df
#ndf[, scols] <- t(t(ndf[, scols]) / rowSums(ndf[, scols]))
# Melt data frame:
tdf <- df %>% gather(key='sample', value='norm_count',-gene_id, -transcript_id, -p_gene, -p_transcript)
# add some grouping metadata
tdf$group <- coldata[match(tdf$sample, coldata$sample_id),]$condition
# change column names for df to prettify reporting
newnames <- rownames(studyDesign)[match(colnames(df)[which(colnames(df) %in% samples)], samples)]
colnames(df)[which(colnames(df) %in% samples)] <- newnames

# Filter for significant genes:



plotDTU <- function(geneId) {
  gdf<-tdf[which(tdf$gene_id==geneId),]
  boxplot <- ggplot(gdf, aes(x=transcript_id, y=norm_count)) + geom_boxplot(aes(fill=group), position="dodge") + geom_dotplot(binaxis="y", stackdir="center", dotsize=0.6, aes(fill=group), position="dodge")+ theme(axis.text.x = element_text(angle = 90, hjust = 1)) +  ylab("Normalised transcript read count") + labs(title=paste("Boxplots showing normalised transcript expression\nlevel across conditions for gene",geneId)) + scale_fill_brewer(palette="Paired")
  print(boxplot)
}

tabulateDTU <- function(geneId) {
  myTable <- knitr::kable(df[which(df$gene_id == geneId),], booktabs=TRUE, table.envir='table*', linesep="") %>% kable_styling(latex_options=c("hold_position", font_size=10))
  print(myTable)
}


#selectedGenes <- unique(as.character(tdf[which(tdf$p_gene < config$adjPValueThreshold),]$gene_id))
selectedGenes <- c("ENSG00000142937", "ENSG00000159377")

for(gene in selectedGenes){
  cat(paste("\n\n## Gene",gene,"\n\n"))
  tabulateDTU(gene)
  
  cat(paste("\n\n", "*The table above shows annotated gene (", gene," p=",round(unique(df[which(df$gene_id == gene),"p_gene"]), digits=5),") and its isoform transcripts. The number of reads mapped to each transcript is shown along with the FDR corrected p-values for DGE and DTU. The read counts are the raw counts. The figure below presents these data in a boxplot format - the read counts are grouped by experimental condition*", sep=""))

  cat("\n\n")
  plotDTU(gene) 
}

```

# Output

The complete transcript mapping data is provided as a CSV format file at
~~~
`r csvExpressedTranscripts`
~~~

An CSV format file containing the transcript mapping counts and parental gene assignments has been prepared and is available in the following folder. 

~~~
`r csvExpressedGenesTranscripts`
~~~

The filtered gene counts and gene isoform counts have been written to Excel format spreadsheet files in the paths indicated below.

~~~
`r xlsExpressedGeneCounts`
`r xlsExpressedFeatureCounts`
~~~


The filtered gene counts and gene isoform counts have been written to Excel format spreadsheet files in the paths indicated above.
~~~
`r xlsExpressedGeneCounts`
`r xlsExpressedFeatureCounts`
~~~


```{r}

```


# Reproducible research - produce your own report

This report has been created using **`Rmarkdown`**, publicly available **`R`** packages, and the \LaTeX document typesetting software for reproducibility. For clarity the **`R`** packages used, and their versions, is listed below.

\fontsize{8}{12}

```{r sessionInfo, eval=TRUE, echo=FALSE, comment=NA}
options(width = 100)
utils:::print.sessionInfo(sessionInfo()[-7], locale=FALSE)
```


```{r, echo=FALSE, include=FALSE, warning="hide"}

```

\fontsize{10}{14}


It is also worth recording the versions of the software that have been used for the analysis.

\fontsize{8}{12}

```{r, engine='bash', echo=FALSE, comment=NA, eval=FALSE}
conda list "salmon|samtools|minimap2|snakemake|rstudio|openjdk"
```


