---
title: "RNASeqvsIsoSeq Functions"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
``` 

Objective: To tabulate the number of full-length reads obtained per gene from Isoseq and order genes from high to low, for comparison with RNAseq data for exact sample

Rationale: To evaluate whether Isoseq output comparable to RNAseq output

Analysis: 1) Downloaded raw subread.bam file from Sequel output 2) CCS and Isoseq3 command line (Lima, Cluster, Polish) 3) Mapped to mouse genome using GMAP 4) Tofu Cupcake 5) Sqanti for isoform characterisation

```{r packages}
library(stringr) # Extract strings
library(dplyr)   # dataframe splitting
library("ggpubr") # correlation plot
```
## IsoSeq Preparation
# Step 1) Annotate2Abundance
Define function for Importing and Merging SQANTI classification file and TOFU abundance file

**Input**: Sqanti_Filter Classification Output file
  * All details of HQ-unique isoforms classified by assigning PacBio output gene Cluster ID to mouse gene name 
**Input**: ToFU Abundance Output file
  * Quantification of number of Full_Length per PacBio_ID
**Output**: Merged txt file by PacBio ID 
  * Merged txt file has the gene name by which the isoform belongs to (as identified by SQANTI) and the quantification of FL_counts (as quantified in TOFU)
  
```{r Annotate2Abudance, echo=TRUE}
SQANTI_input <- function(sample){
  ## Define directory input
  sqanti_dir <- "/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/WholeTranscriptome/Individual/SQANTI2"
  
  ## List all files in SQANTI directory 
  #print("All Classification files from SQANTI")
  all_classification_files <- list.files(path = sqanti_dir , pattern = ".collapsed.filtered.rep_classification.filtered_lite_classification.txt", full.names = TRUE)
  #print(all_classification_files)
  #return(classification.files)
  
  ## Grep specific file based on sample name
  print(paste("Input SQANTI Filter output file for Sample", sample))
  classification_file <- grep(sample, all_classification_files, fixed = TRUE, value = TRUE)
  print(classification_file)
  
  ## Read specific sample file
  classification_dat <- read.delim2(file = classification_file, header = TRUE)
  classification_dat <<- data.frame(classification_dat)
  
  print(paste("SQANTI Classification file of Sample", sample))
  return(head(classification_dat))
}

TOFU_input <- function(sample){
  ## Define directory input
  tofu_dir <- "/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/WholeTranscriptome/Individual/ToFU"
  
  ## List all files in TOFU directory
  #print("All Abundance files from TOFU")
  all_abundance_files <- list.files(path = tofu_dir , pattern = ".collapsed.filtered.abundance.txt", full.names = TRUE)
  #print(all_abundance_files)
  
  ## Grep specific file based on sample name
  print(paste("Input SQANTI Filter output file for Sample", sample))
  abundance_file <- grep(sample, all_abundance_files, fixed = TRUE, value = TRUE)
  print(abundance_file)
  
  ## Read specific sample file
  abundance_dat <- read.table(file = abundance_file, header = TRUE)[-c(1:14),] #remove rows 1:14 of descriptor
  abundance_dat <<- data.frame(abundance_dat)
  colnames(abundance_dat) <- c("PacBio_Id","count_fl","count_nfl","count_nfl_amb","norm_fl","norm_nfl","norm_nfl_amb")
  
  # View specific sample file
  print(paste("Abundance file of Sample", sample))
  return(head(abundance_dat))
}

Annotate2Abundance_IsoSeq <- function(sample,classification_dat,abundance_dat){
  # Merge SQANTI_Classssification file for genename with TOFU_Abundance file for FL counts
  Merge_IsoSeq <- merge(classification_dat,abundance_dat,by.x = "isoform", by.y = "PacBio_Id")
  #write.csv(Merge, file = print(paste0(sample,"Merged.csv")))
  
  # save Merge file to workspace: https://stackoverflow.com/questions/32563153/function-save-returned-data-frame-to-workspace
  print(paste("Merged file of SQANTI Classification and Abundance File of Sample", sample))
  Merge_IsoSeq <<- Merge_IsoSeq
  return(Merge_IsoSeq)
}
```
# Step 2)SumFLCounts
Define function that the FL Counts for all transcripts per gene 

**Motivation**: SQANTI Filter classification outputs one gene with multiple isoforms, thus complicates correlation with RNA-Seq Gene Expression Counts. 
PacBio FL count is presented per isoform rather than per gene. However, FeatureCount's output from RNA-seq data is on a gene level. Therefore FL counts from IsoSeq needs to be summed for more convenient comparison: Total FL Counts of Transcripts per Gene from IsoSeq vs Raw Gene Counts from RNASeq 

Alternative option: select only isoform with the highest number of FL counts, yet biased results especially given if many isoforms with similar or slgihtly smaller number of FL-counts. 
Assumptions: RNA-seq captures expression of all RNA transcripts irrespective of isoforms

```{r}
#  %>% group_by(Genes) %>% = group/splice MCount dataframe by genes
# In the groups, keep RNAseq Counts and PacBio ID columns as they are 
# sum(PacBio_FL_Counts)
SumFLCounts <- function(dat){
  # input dat dataframe, group by "associated_gene" column (i.e. group by genes), and include 2 colums:
  # column 1 = labelled "PacBio Isoform" == isoform column; note [1] for referring to entire column
  # column 2 = labelled "PacBio FL Counts" == sum of "count_Fl" column
  
  Merge_IsoSeq_SumFL <- dat %>% group_by(associated_gene) %>%
    summarize(PacBio_Isoform = isoform[1], PacBio_FL_Counts = sum(count_fl))
  
  Merge_IsoSeq_SumFL$associated_gene <- as.character(Merge_IsoSeq_SumFL$associated_gene)
  return(head(Merge_IsoSeq_SumFL))
  
  # Validation that the function works 
  print("Validation of summing PacBio FL")
  print("Original input data from ToFU Abundance files for the Arf3 gene")
  print(dat[which(dat$associated_gene == "Arf3"),c(7,40)]) #only print column 7 and 40 = "associated_gene" and "fl_count"
  print("Summed PacBio FL count for the Arf3 gene, saved as new dataframe for downstream analysis")
  Merge_IsoSeq_SumFL[which(Merge_IsoSeq_SumFL$associated_gene == "Arf3"),]
}

Validation_SumFLCounts <- function(gene,dat){
  # Validation that the function works 
  print(paste("Validation of summing PacBio FL"))
  print(paste("Original input data from ToFU Abundance files for the Gene", gene))
  df <- dat[which(dat$associated_gene == gene),c(7,40)] #only print column 7 and 40 = "associated_gene" and "fl_count"
  
  print(paste("Summed PacBio FL count for the Gene", gene , "saved as new dataframe for downstream analysis"))
  # Merge_IsoSeq_SumFL is the output dataframe from SumFLCounts Function
  df2<- Merge_IsoSeq_SumFL[which(Merge_IsoSeq_SumFL$associated_gene == gene),] 
  
  validation <- list(df,df2)
  validation <<- validation
  return(validation)
}
```
## Step 3)RNASeq Preparation

**Input**: FeatureCounts of all RNASeq samples (STAR Aligned to mm10 genome, and annotated to Gencode Mouse V20 gtf file) at gene level 
**Output**: FeatureCount of specific sample 

```{r, echo = TRUE}
Input_RNAseq <- function(sample){
  # Define directory input 
  featurecount_dir <- "/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/WholeTranscriptome/WT8_ISOSEQ/Post/FeatureCounts"
  
  print("Input FeatureCount for All Samples")
  # Import featurecount file with annotated mgisymbol names
  featurecount_file <- list.files(path = featurecount_dir, pattern = "Final_WT8_genenames.mx", full.names = TRUE) 
  featurecount_dat <- read.delim2(featurecount_file, row.names = 1, header = TRUE)
  # Featurecount_dat <<- featurecount_dat #save to workspace
  print(head(featurecount_dat))
  
  # Grep specific sample column with counts and ensembl_name (column 1)
  RNASeq <- data.frame(featurecount_dat[,sample],row.names = (rownames(featurecount_dat)))
  colnames(RNASeq)[1] <- paste("RNASeq",sample,"Raw Counts")
  print(paste("Input FeatureCount for Sample", sample))
  #print(head(RNASeq))
  
  # To extract the mgi_symbol separated by _ per row (carried out with a for loop)
  Genes <- data.frame(lapply(rownames(RNASeq), function(x){gene<-word(x,2, sep = fixed ('_'))}))
  Genes <- data.frame(t(Genes))
  colnames(Genes)[1] <- "Mgi_Symbol"
  RNASeq$Mgi_Symbol <- as.character(RNASeq$Mgi_Symbol)
  
  RNASeq <- cbind(RNASeq,Genes)
  RNASeq <<- RNASeq
  return(head(RNASeq))
}

```
## Step 4) Merge RNASeq and IsoSeq

**Input**: Sample-specific Isoseq and RNASeq Counts
**Output**: Dataframe "Full_Merge": Merged Counts across IsoSeq and RNASeq by gene names

Also to call out specific counts of AD-associated genes, created function AD_Counts. Genes of interest specified in working script. 

```{r}
Merge_RNASeq_Isoseq <- function(IsoSeq,RNASeq){
  Full_Merge <- merge(IsoSeq,RNASeq,by.x = "associated_gene",by.y = "Mgi_Symbol",all = TRUE)
  Full_Merge <<- Full_Merge
  return(head(Full_Merge))
}

AD_Counts <- function(AD_Genes){
  
  # Create new dataframe for appending output from loop
  AD_Genes_Merged <- data.frame() 
  
  for(i in AD_Genes){
    df <- Full_Merge[which(Full_Merge$associated_gene == i),]
    AD_Genes_Merged <- rbind(AD_Genes_Merged, df)
  }
  
  return(AD_Genes_Merged)
}
```
## Step 5) Data Review for Full_Merge: RNASeq vs IsoSeq

**Motivation**: Within Full_Merge dataframe, interested to know which genes are detected only by IsoSeq, only by RNASeq, and alone. Also later downstream, able to plot the number of respective counts for these genes. 

```{r}
Missing_Reads_Review <- function(){
  print(paste("Total Number of Genes in Full_Merge of IsoSeq and RNASeq:", dim(Full_Merge)[1]))
  
  # Create "Detection column" in Full_Merge depending on whether reads present/absent in technology
  colnames(Full_Merge)
  colnames(Full_Merge) <- c("associated_gene","PacBio_Isoform_ID","Isoseq_FL_Counts","RNASeq_Raw_Counts") #double check col names
  
  Full_Merge$Detection_Isoseq = ifelse(is.na(Full_Merge$Isoseq_FL_Counts),"noIsoSeq","IsoSeq")
  Full_Merge$Detection_RNAseq = ifelse(is.na(Full_Merge$RNASeq_Raw_Counts),"noRNASeq","RNASeq") 
  Full_Merge$Detection = paste(Full_Merge$Detection_Isoseq,Full_Merge$Detection_RNAseq,sep="_") #concatenate columns
  
  # Check factor level for detection
  Full_Merge$Detection <- as.factor(Full_Merge$Detection)
  # print(levels(Full_Merge$Detection))
  
  print(paste("Total Number of Genes Detected in IsoSeq AND RNASeq:",length(which(Full_Merge$Detection == "IsoSeq_RNASeq"))))
  print(paste("Total Number of Genes Detected in IsoSeq but not RNASeq:",length(which(Full_Merge$Detection == "IsoSeq_noRNASeq"))))
  print(paste("Total Number of Genes Detected in RNASeq but not IsoSeq:",length(which(Full_Merge$Detection == "noIsoSeq_RNASeq"))))
  
  # Substitute NA values for 0 to plot correlation
  Full_Merge$Isoseq_FL_Counts[is.na(Full_Merge$Isoseq_FL_Counts)] <- 0
  Full_Merge$RNASeq_Raw_Counts[is.na(Full_Merge$RNASeq_Raw_Counts)] <- 0
  
  Full_Merge <<- Full_Merge #important to update df 
  return(head(Full_Merge))
}
```
## Step 6) Correlation

**output**: Correlation of Gene Expression of IsoSeq FL Counts vs RNASeq Raw Counts. Correlation coefficient calculated from pearson's method (assuming parametric) and considers 

```{r}
Run_Corplot <- function(dat){
  Corplot <- ggscatter(dat, y = "RNASeq_Raw_Counts", x = "Isoseq_FL_Counts", 
            color = "Detection",
            palette = c("red", "black"," blue"),
            add = "reg.line", conf.int = TRUE, 
            cor.coef = TRUE, cor.method = "pearson",
            ylab = "RNASeq_Raw_Counts", xlab = "PacBio_FL_Counts")
  
  Corplot + labs(fill = "Dose (mg)")
  
  Corplot
}

Run_Corplot(Full_Merge)
```
## Step 7) Missing Reads

**input**: Genes either detected by IsoSeq or RNASeq from Full_Merge dataframe (IsoSeq and RNASeq Counts/gene) 
**output**: Plot of those genes with its respective counts
```{r}
Missing_Reads_Plot <- function(dat){
  # Only genes with either IsoSeq/RNASeq Reads
  Missing <- dat[which(dat$Detection == "noIsoSeq_RNASeq" | dat$Detection == "IsoSeq_noRNASeq" ),]
  
  Missing_Plot <- ggscatter(Missing, y = "RNASeq_Raw_Counts", x = "Isoseq_FL_Counts", 
                            color = "Detection",
                            label = "name", repel = TRUE,
                            ylab = "RNASeq_Raw_Counts", xlab = "PacBio_FL_Counts")
  
  Missing_Plot_2 <- Missing_Plot + scale_color_discrete(name ="Detection", 
                                labels=c("IsoSeq only", "RNASeq only"))
  
  Missing <<- Missing # for downstream use
  return(Missing_Plot_2)
}


Missing_Genes <- function(value){
  print(paste("Genes with no IsoSeq Reads but RNASeq RawCounts >", value))
  print(Missing[which(Missing$RNASeq_Raw_Counts > value),1])
  print("Genes with only IsoSeq Reads, and no RNASeq Reads")
  print(Missing[which(Missing$Detection == "IsoSeq_noRNASeq"),1])
}

```

